src/coreclr/tools/Common/TypeSystem/IL/UnsafeAccessors.cs

```diff
@@ -0,0 +1,377 @@
+                : targetTypeMaybe;
+
+            // Due to how some types degrade, we block on parameterized
+            // types. For example ref or pointer.
+            if ((targetType.IsParameterizedType && !targetType.IsArray)
+                || targetType.IsFunctionPointer)
+            {
+                ThrowHelper.ThrowBadImageFormatException(InvalidUnsafeAccessorUsage);
+            }
+
+            return targetType;
+        }
+
+        private static bool TrySetTargetMethod(ref GenerationContext context, string name)
+        {
+            TypeDesc targetType = context.TargetType;
+
+            TypeDesc declTypeDesc;
+            TypeDesc maybeTypeDesc;
+            foreach (MethodDesc md in targetType.GetMethods())
+            {
+                // Check the target and current method match static/instance state.
+                if (context.IsTargetStatic != md.Signature.IsStatic)
+                {
+                    continue;
+                }
+
+                // Check for matching name
+                if (!md.Name.Equals(name))
+                {
+                    continue;
+                }
+
+                // Validate calling convention.
+                if ((MethodSignatureFlags.UnmanagedCallingConventionMask & md.Signature.Flags)
+                    != (MethodSignatureFlags.UnmanagedCallingConventionMask & context.Declaration.Signature.Flags))
+                {
+                    continue;
+                }
+
+                // Validate the return type and prepare for validating
+                // the current signature's argument list.
+                int sigCount = context.Declaration.Signature.Length;
+                if (context.Kind == UnsafeAccessorKind.Constructor)
+                {
+                    if (!md.Signature.ReturnType.IsVoid)
+                    {
+                        continue;
+                    }
+                }
+                else
+                {
+                    declTypeDesc = context.Declaration.Signature.ReturnType;
+                    maybeTypeDesc = md.Signature.ReturnType;
+                    if (declTypeDesc != maybeTypeDesc)
+                    {
+                        continue;
+                    }
+
+                    // Non-constructor accessors skip the first argument
+                    // when validating the target argument list.
+                    sigCount--;
+                }
+
+                // Validate argument count matches.
+                if (sigCount != md.Signature.Length)
+                {
+                    continue;
+                }
+
+                // Validate arguments match - reverse order
+                for (; sigCount > 0; --sigCount)
+                {
+                    declTypeDesc = context.Declaration.Signature[sigCount];
+                    maybeTypeDesc = md.Signature[sigCount];
+                    if (declTypeDesc != maybeTypeDesc)
```

Comment:
This comparison will not disambiguate between:

```csharp
unsafe class Blah
{
    static void Method(delegate* unmanaged[Stdcall]<void> param) { }
    static void Method(delegate* unmanaged[Cdecl]<void> param) { }
}
```

Per the ECMA-335 change we did for function pointers, the only thing that matters for function pointer identity is whether the calling convention is unmanaged. These two signatures look indistinguishable.

Instead of doing a foreach over all methods and replicating the signature matching rules, can we just call the `TypeDesc.GetMethod` overload that takes a `MethodSignature` we took from the unsafe accessor method? (Same comment to the code that that locates fields using a foreach).

Reply:
> This comparison will not disambiguate between:

That is true. We aren't interested in comparing modopts and modreqs for this feature.

Reply:
> > This comparison will not disambiguate between:
> 
> That is true. We aren't interested in comparing modopts and modreqs for this feature.

Reading through the issue, I understand we need to skip them to be able to refer to things like IsExternalInit. But also not being able to distinguish overloads that are expressible in C# today doesn't give me warm and fuzzy feelings either. Ever since the Roslyn team decided to forget that modops are not CLS compliant, they started showing up for more and more patterns. I'm worried that 5 years from now we'll have even more cases of "this calls an arbitrary overload instead of the one I obviously wanted".

Reply:
We could potentially encode the modifiers into custom attributes on the unsafe accessor, but I can see how that would make things _very complicated_ on the CoreCLR side.

I guess we'll have to live with this although it might be nicer to error out if the ignored modifiers make things ambiguous instead of picking the first thing we found.

Reply:
> I guess we'll have to live with this although it might be nicer to error out if the ignored modifiers make things ambiguous instead of picking the first thing we found.

I'm open to ideas on how to do that without parsing everything multiple times. Is there an API that helps with that. @davidwrighton brought up a similar concern offline.

Reply:
My 2 cents is that not being able to specify custom modifiers will end up causing problems (which I commented in the original issue, but hadn't thought of something like function pointers at the time).
I think for now, since we aren't going to support specifying custom modifiers, we should throw an exception on ambiguity. If there was an overload with the exact same modifiers as those in our method declaration, I also think it would be good if it didn't throw on ambiguity in that case, and instead just used the matching overload.